clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractMembershipManager.java

```diff
@@ -1129,6 +1168,12 @@ private CompletableFuture<Void> assignPartitions(
             if (exception == null) {
                 // Enable newly added partitions to start fetching and updating positions for them.
                 subscriptions.enablePartitionsAwaitingCallback(addedPartitions);
+
+                Set<TopicPartition> allAssignedPartitions = assignedPartitions.stream()
+                    .map(tip -> new TopicPartition(tip.topic(), tip.partition()))
+                    .collect(Collectors.toSet());
+
+                notifyAssignmentChange(allAssignedPartitions);
```

Comment:
uhm this should probably go right after the subscription is updated, not after the callback complete. Having it here, I guess it would allow for a race if there is a call in the app thread to consumer.close, while a reconciliation is running this code. If the call to consumer.close happens after callbacks complete on ln 1167 above, but before making it to this line, the app thread could potentially close revoking partitions without including the ones just added here right? 

So we could just call this notify right after `updateSubscriptionAwaitingCallback` (which is the one that does the actual update of the subscription state). From that moment on the partitions are effectively assigned even though the onPartitionsAssigned may have not completed.

Reply:
Moved `notifyAssignmentChange()` to right after `subscriptions.assignFromSubscribedAwaitingCallback()` in `updateSubscriptionAwaitingCallback()`.